src/Storages/MergeTree/EphemeralLockInZooKeeper.h

```diff
@@ -86,14 +88,14 @@ class EphemeralLockInZooKeeper : public boost::noncopyable
     ~EphemeralLockInZooKeeper();
 
 private:
-    zkutil::ZooKeeper * zookeeper = nullptr;
+    KeeperAccessPtr zookeeper;
     String path_prefix;
     String path;
     String holder_path;
 };
 
 std::optional<EphemeralLockInZooKeeper> createEphemeralLockInZooKeeper(
-    const String & path_prefix_, const String & temp_path, zkutil::ZooKeeper & zookeeper_, const String & deduplication_path);
+    const String & path_prefix_, const String & temp_path, const KeeperAccessPtr & zookeeper_, const String & deduplication_path);
```

Comment:
`ZooKeeper` client is not created on insert execution path. So, we have no control over its creation, → just using it.

Reply:
Hm, yes, but we can wrap real `zkutil::ZooKeeper` client into an object that implements the same interface and calls methods of wrapped client. We will not need to replace `zkutil::ZooKeeper` with `KeeperAccessPtr` everywhere. However, it will require to make all methods virtual (and make an interface) and wrap all of them (a lot of boilerplate). So I'm not sure which approach is better, maybe `KeeperAccessPtr` is okay.

Reply:
Another (questionable and quite dirty) approach: introduce a thread_local pointer to a struct with fault injection settings (or put it into `ThreadStatus`). Create an object that stores these settings, sets the pointer in ctor and resets it in dtor. Check the pointer in `ZooKeeper::pushRequest` (before `try` to avoid finalization) and throw exception if needed. 

But I do not insist on any these approaches, feel free to keep `KeeperAccessPtr`.